File: problem_036.c
// -*- compile-command: "gcc -o problem_036 problem_036.c -Wall" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <stdbool.h>

bool decIsPalindromic(int input)
{
    int decompose = input;
    int reverse = 0;

    while (decompose > 0)
    {
        reverse *= 10;
        reverse += decompose %10;
        decompose /= 10;
    }

    if ( reverse == input )
        return true;
    return false;
}

bool binIsPalindromic(int *numberIn, int length)
{
    int index = 0;
    int numStart = 0;

    for (numStart=0 ; numberIn[numStart]==0 ; numStart++);
    for (index=numStart ; index<length ; index++)
        if (numberIn[index] != numberIn[length-1-(index-numStart)])
            return false;
    return true;
}

int main()
{
    printf("Project Euler - Problem 36:\n"
           "Find the sum of all numbers, less than one million, which are palindromic in base 10 and base 2.\n\n");

    const int lenNum = 20;
    int binNum[lenNum];
    int index, number;
    int sum = 0;

    //Clear our binnum
    for (index=0 ; index<lenNum ; index++)
        binNum[index]=0;

    for (number=1 ; number<1000000 ; number++)
    {
        //Increment array to match count
        binNum[lenNum-1] += 1;

        //Carry-over each element if needed
        for (index=lenNum-1 ; index>0 ; index--)
        {
            if (binNum[index] > 1)
            {
                binNum[index-1] += 1;
                binNum[index]   =  0;
            }
        }

        if (decIsPalindromic(number) == true && binIsPalindromic(binNum, lenNum) == true)
        {
            sum += number;

            /*
            //Debug
            printf("Palindrome found: %d; ", number);
            for (index=0 ; index<lenNum ; index++)
                printf("%d", binNum[index]);
            printf("\n");
            */
        }
    }

    printf("Sum: %d\n", sum);

    return 0;
}

File: problem_036.rs
// -*- compile-command: "rustc -o problem_036_rs problem_036.rs" -*-
// Copyright (c) 2016 Michael Caldwell
//use std::num;

fn is_palindromic_bin(mut n: u64) -> bool {
    let mut bin = [0; 20];

    // Convert into binary number
    for i in (0..20).rev() {
        let b = (2 as f64).powi(i) as u64;
        if n >= b {
            n -= b;
            bin[i as usize] = 1;
        }
    }

    // Find end of BIN number, then compare with other side
    let mut compare: i32 = -1;
    for i in (0..20).rev() {
        if bin[i] == 1 && compare == -1{
            compare = 0;
        }
        if compare > -1 {
            if bin[i] != bin[compare as usize] {
                return false;
            }
            compare += 1;
        }
    }

    return true;
}

fn reverse(mut n: u64) -> u64 {
    let mut o = 0;

    while n > 0 {
        o *= 10;
        o += n % 10;
        n /= 10;
    }

    return o;
}


fn main() {
    println!("Project Euler - Problem 36");
    println!("Find the sum of all numbers, less than one million, which are palindromic in base 10 and base 2.\n");

    let mut sum = 0;

    for n in 1..1000000 {
        if n == reverse(n) {
            if is_palindromic_bin(n) {
                sum += n;
            }
        }
    }

    println!("{}", sum);
}


//872187
